src/doc/unstable-book/src/language-features/c-ffi-const.md

```diff
@@ -0,0 +1,51 @@
+# `c_ffi_const`
+
+The `#[c_ffi_const]` attribute applies clang's `const` attribute to foreign
+functions declarations. 
+
+That is, `#[c_ffi_const]` functions shall have no effects except for its return
+value, which can only depend on the values of the function parameters, and is
+not affected by changes to the observable state of the program.
+
+The behavior of calling a `#[c_ffi_const]` function that violates these
+requirements is undefined.
+
+This attribute enables Rust to perform common optimizations, like sub-expression
+elimination, and it can avoid emitting some calls in repeated invocations of the
+function with the same argument values regardless of other operations being
+performed in between these functions calls (as opposed to `#[c_ffi_pure]`
+functions).
+
+## Pitfalls
+
+A `#[c_ffi_const]` function can only read global memory that would not affect
+its return value for the whole execution of the program (e.g. immutable global
+memory). `#[c_ffi_const]` functions are referentially-transparent and therefore
+more strict than `#[c_ffi_pure]` functions.
+
+A common pitfall involves applying the `#[c_ffi_const]` attribute to a
+function that reads memory through pointer arguments which do not necessarily
+point to immutable global memory.
+
+A `#[c_ffi_const]` function that returns unit has no effect on the abstract
+machine's state, and a `#[c_ffi_const]` function cannot be `#[c_ffi_pure]`.
+
+A diverging and C or C++ `const` function is unreachable. Diverging via a
+side-effect (e.g. a call to `abort`) violates `const` pre-conditions. Divergence
+without side-effects is undefined behavior in C++ and not possible in C. In C++,
+the behavior of infinite loops without side-effects is undefined, while in C
+these loops can be assumed to terminate. This would cause a diverging function
+to return, invoking undefined behavior.
```

Comment:
This paragraph confuses me, I think it might incidentally mislead users too, and I don't even think it is technically correct

- C's "assumed to terminate" is indistinguishable from "UB if non-terminating", so that last sentence is superfluous (and communicates a wrong picture of UB)
- C doesn't actually say this about all infinite loops, it has an exception carved out for loops controlled by constant expressions, such as `while (1) { ... }`

I think we should just say as a separate axiom that functions marked with this attribute must not diverge. GCC's docs don't say this as far as I can see (at least explicitly -- arguably non-termination is an observable effect), but it's kind of the safer assumption because who knows what LLVM's implicit assumptions around `readnone`/`readonly` functions are at the moment. It also seems useful to assume termination of const/pure functions to be able to execute them speculatively.